### PR TITLE
add pslib(pserver) to paddle, an industrial scale high performance parameter server library

### PR DESCRIPTION
**Pull request by colourful-tree**

### CODE REVIEW

File: paddle/fluid/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 add_subdirectory(memory)
 add_subdirectory(platform)
 add_subdirectory(framework)
+#add_subdirectory(distributed)
```
Comment (Yancey0623) [inline paddle/fluid/CMakeLists.txt]: why comment on this directory?

File: python/paddle/fluid/distributed/ps_pb2.py [hunk 2]
```diff
@@ -0,0 +1,1506 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (Yancey0623) [inline python/paddle/fluid/distributed/ps_pb2.py:13]: Please don't add the pb2 file into the repo, maybe compile the protobuf by cmake is a better way, for example: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/CMakeLists.txt#L36 .

File: paddle/fluid/CMakeLists.txt (hunk 1 above)
Comment (colourful-tree) [reply paddle/fluid/CMakeLists.txt]: because we don't use this, had delete this.

File: python/paddle/fluid/distributed/ps_pb2.py (hunk 2 above)
Comment (colourful-tree) [reply python/paddle/fluid/distributed/ps_pb2.py:13]: the protobuf is related to pslib, now, don't want to exposure of this.

File: python/paddle/fluid/distributed/node.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        table.accessor.accessor_class = "DownpourFeatureValueAccessor"
+        table.accessor.sparse_sgd_param.learning_rate = learning_rate
+        table.accessor.sparse_sgd_param.initial_g2sum = 3
+        table.accessor.sparse_sgd_param.initial_range = 1e-4
+        table.accessor.sparse_sgd_param.weight_bounds.extend([-10, 10])
+
+        table.accessor.embedx_dim = 8
+        table.accessor.embedx_threshold = 5
+        table.accessor.fea_dim = 11
+        #table.accessor.fea_dim = abs(reduce(lambda x, y: x * y,
```
Comment (guru4elephant) [inline python/paddle/fluid/distributed/node.py]: remove this comment

File: python/paddle/fluid/distributed/node.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            window (int): push params frequency
+            worker: it is pslib.DownpourTrainerParameter 
+        Examples:
+            worker = DownpourWorker(1)
+    """
+
+    def __init__(self, window):
+        self.window = window
+        self.worker_ = pslib.DownpourTrainerParameter()
+        #self.worker_.pull_dense_per_batch = window
```
Comment (guru4elephant) [inline python/paddle/fluid/distributed/node.py]: remove this comment

File: python/paddle/fluid/distributed/ps_instance.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                if 0 == self._rankid % self._proc_per_node % 2:
+                    self._node_type = 0
+                else:
+                    self._node_type = 1
+            else:
+                self._node_type = -1
+        else:
+            self._node_type = -1
+
+        #if self._rankid == 0:
```
Comment (guru4elephant) [inline python/paddle/fluid/distributed/ps_instance.py]: remove this comment

**Review by reyoung — DISMISSED** [review]
LG API

File: paddle/fluid/framework/CMakeLists.txt [hunk 6]
```diff
@@ -89,8 +89,8 @@ nv_test(data_device_transform_test SRCS data_device_transform_test.cu
 
 if(WITH_GPU)
   if (WIN32)
-    # windows treat symbolic file as a real file, which is different with unix
-    # We create a hidden file and compile it instead of origin source file.
+#windows treat symbolic file as a real file, which is different with unix
```
Comment (typhoonzero) [inline paddle/fluid/framework/CMakeLists.txt:95]: comment style should not be changed

File: python/paddle/fluid/distributed/helper.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                 hadoop_bin=""):
+        assert user != None
+        assert passwd != None
+        assert hadoop_bin != None
+        import ps_pb2 as pslib
+        self.fs_client = pslib.FsClientParameter()
+        #if fs_type == "afs":
+        #    fs_client.fs_type = pslib.FsApiType.AFS
+        #else:
+        #    fs_client.fs_type = pslib.FsApiType.HDFS
```
Comment (typhoonzero) [inline python/paddle/fluid/distributed/helper.py]: no commented out code should be checked in.

File: python/paddle/fluid/distributed/helper.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        user (string): hadoop param
+        passwd (string): hadoop param
+        hadoop bin (string): hadoop param
+    Examples:
+        fs = FileSystm()
+    """
+
+    def __init__(self,
+                 fs_type="afs",
+                 uri="afs://tianqi.afs.baidu.com:9902",
```
Comment (typhoonzero) [inline python/paddle/fluid/distributed/helper.py]: need clean up urls for internal usage

File: python/paddle/fluid/distributed/helper.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (typhoonzero) [inline python/paddle/fluid/distributed/helper.py:1]: here's one more space after '#'

File: python/paddle/fluid/distributed/ps_pb2.py [hunk 10]
```diff
@@ -0,0 +1,2296 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (typhoonzero) [inline python/paddle/fluid/distributed/ps_pb2.py:1]: Do not commit generated _pb2 files.

**Review by typhoonzero — DISMISSED** [review]

File: cmake/external/libmct.cmake [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ENDIF()
+
+INCLUDE(ExternalProject)
+
+SET(LIBMCT_PROJECT       "extern_libmct")
+IF((NOT DEFINED LIBMCT_VER) OR (NOT DEFINED LIBMCT_URL))
+  MESSAGE(STATUS "use pre defined download url")
+  SET(LIBMCT_VER "0.1.0" CACHE STRING "" FORCE)
+  SET(LIBMCT_NAME "libmct" CACHE STRING "" FORCE)
+  SET(LIBMCT_URL "https://raw.githubusercontent.com/PaddlePaddle/Fleet/release/${LIBMCT_VER}/${LIBMCT_NAME}.tar.gz" CACHE STRING "" FORCE)
```
Comment (typhoonzero) [inline cmake/external/libmct.cmake:34]: might need to change to use baidu bos storage later.

**Review by typhoonzero — APPROVED** [review]

**Review by reyoung — APPROVED** [review]
LG API

File: python/paddle/fluid/distributed/ps_pb2.py [hunk 12]
```diff
@@ -0,0 +1,2296 @@
...
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
```
Comment (jacquesqiao) [inline python/paddle/fluid/distributed/ps_pb2.py:8]: It's strange to put protobuf compiled result into code base.

**Pull request merged**